File: hono-bun/src/core/pagination.ts
import { z } from "@hono/zod-openapi";

export const PaginationQuerySchema = z.object({
  limit: z.coerce.number().min(1).max(100).optional().default(20).openapi({
    description: "Number of items per page",
    example: 20,
  }),
  offset: z.coerce.number().min(0).optional().default(0).openapi({
    description: "Number of items to skip",
    example: 0,
  }),
});

export const PaginationMetaSchema = z.object({
  total: z.number().openapi({ description: "Total item count" }),
  limit: z.number().openapi({ description: "Items per page" }),
  offset: z.number().openapi({ description: "Items skipped" }),
});

export type PaginationQuery = z.infer<typeof PaginationQuerySchema>;
export type PaginationMeta = z.infer<typeof PaginationMetaSchema>;

export function buildPaginationMeta(
  total: number,
  limit: number,
  offset: number,
): PaginationMeta {
  return { total, limit, offset };
}

File: fastapi/src/core/pagination.py
from pydantic import BaseModel, Field


class PaginationQuery(BaseModel):
    limit: int = Field(default=20, ge=1, le=100, description="Number of items per page")
    offset: int = Field(default=0, ge=0, description="Number of items to skip")


class PaginationMeta(BaseModel):
    total: int
    limit: int
    offset: int


def build_pagination_meta(total: int, limit: int, offset: int) -> PaginationMeta:
    return PaginationMeta(total=total, limit=limit, offset=offset)
